src/NuGet.Core/NuGet.Packaging/Rules/IconAndIconUrlUsageWarning.cs

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using NuGet.Common;
+
+namespace NuGet.Packaging.Rules
+{
+    public class IconAndIconUrlUsageWarning : IPackageRule
+    {
+        public string MessageFormat { get; }
+
+        public IconAndIconUrlUsageWarning(string messageFormat)
+        {
+            MessageFormat = messageFormat ?? throw new ArgumentNullException(nameof(messageFormat));
+        }
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            var nuspecReader = builder?.NuspecReader;
+            var icon = nuspecReader.GetIcon();
```

Comment:
Will this throw a `NullReferenceException` when `builder` is `null`? Consider something like:

```csharp
var icon = nuspecReader?.GetIcon();
var iconUrl = nuspecReader?.GetIconUrl();

if (!string.IsNullOrEmpty(icon) && !string.IsNullOrEmpty(iconUrl)
{
   ...
{
```

Reply:
Same comment for `IconUrlDeprecationWarning`

Reply:
Good suggestion. I will make the changes.

According to the PackageArchiveReader base [implementation](https://github.com/NuGet/NuGet.Client/blob/4590967639182b183b247355596aedc7d0855a86/src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs#L112), NuspecReader cannot be null, at least inside the NuGet world.